racket/collects/racket/file.rkt

```diff
@@ -134,100 +139,410 @@
+       (bytes->path bs)))
+   #:syntactic-directory-error
+   (λ (result)
+     (raise-arguments-error
+      who
+      (bad-result-msg "syntactic directory path not allowed")
+      "prefix" prefix
+      "suffix" suffix
+      "produced" result))
+   #:complete-with-base-error
+   (λ (result)
+     ;; On Windows, base-dir could be a drive specification,
+     ;; in which case it is ok for result to be an absolute path without a drive.
+     (raise-arguments-error
+      who
+      (bad-result-msg "complete path can not be combined with base-dir")
+      "prefix" prefix
+      "suffix" suffix
+      "produced" result
+      "base-dir" base-dir))))
+
+
+(define-for-syntax (syntax->tmp-context-string stx)
+  (define line   (syntax-line stx))
+  (define col    (syntax-column stx))
+  (define source (syntax-source stx))
+  (define pos    (syntax-position stx))
+  (define str-src
+    (cond [(path? source)
+           (regexp-replace #rx"^<(.*?)>(?=/)"
+                           (path->relative-string/library source)
+                           (lambda (_ s) (string-upcase s)))]
+          [(string? source) source]
+          [else #f]))
+  (define str-loc
+    (cond [(and line col) (format "-~a-~a" line col)]
+          [pos (format "--~a" pos)]
+          [else ""]))
+  (define combined-str (string-append (or str-src "rkttmp") str-loc))
+  (define sanitized-str (regexp-replace* #rx"[<>:\"/\\|]" combined-str "-"))
+  (define max-len 50) ;; must be even
+  (define not-too-long-str
+    (cond [(< max-len (string-length sanitized-str))
+           (string-append (substring sanitized-str 0 (- (/ max-len 2) 2))
+                          "----"
+                          (substring sanitized-str
+                                     (- (string-length sanitized-str)
+                                        (- (/ max-len 2) 2))))]
+          [else sanitized-str]))
+  not-too-long-str)
+
+(define-for-syntax (infer-temporary-file-template stx)
+  (string-append (syntax->tmp-context-string stx) "_~a"))
+
+(define-for-syntax (infer-temporary-file-context-bytes stx)
+  ;; We rely on the sanitization above having been used
+  ;; successfully by Racket from time immemorial.
+  (string->bytes/utf-8 (syntax->tmp-context-string stx)))
+
+(define-syntax (make-temporary-directory stx)
+  (with-syntax ([app (datum->syntax stx #'#%app stx)])
+    (syntax-case stx ()
+      [x
+       (identifier? #'x)
+       #'make-temporary-directory/proc]
+      [(_)
+       #`(app make-temporary-directory/proc
+              '#,(infer-temporary-file-template stx))]
+      [(_ #:base-dir base-dir)
+       (not (keyword? #'base-dir))
+       #`(app make-temporary-directory/proc
+              '#,(infer-temporary-file-template stx)
+              #:base-dir base-dir)]
+      [(_ . whatever)
+       #'(app make-temporary-directory/proc . whatever)])))
+
 (define-syntax (make-temporary-file stx)
   (with-syntax ([app (datum->syntax stx #'#%app stx)])
```

Comment:
I think this (existing) line is actually a no-op, because the [docs](https://docs.racket-lang.org/reference/stxops.html#%28def._%28%28quote._~23~25kernel%29._datum-~3esyntax%29%29) say that "if `v` is already a syntax object, then there is no conversion, and `v` is returned unmodified."

Reply:
Surely it's not a no-op: this looks like it's adjusting the `#%app` syntax object to be in the syntax scope/hygiene (terminology unclear to me) of the original `stx`.

Reply:
That would be true if it were `'#%app` (as I had assumed until this evening), but it is actually `#'#%app`, so, IIUC, it will not.

It seems very possible this line of code was _supposed_ to do that, but I was never clear on exactly _why_ it seemed to be trying to do that, so, if I'm right that it is not actually doing anything now, I'm inclined not to make it start doing anything.